Support LDAP dns longer than 255 characters

come-nc opened the pull request:
Adds an `ldap_full_dn` column to store the dn, and only store a sha256 hash in the `ldap_dn` which is shorter and can be indexed without trouble.
Migration still needs to be implemented.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Support LDAP dns longer than 255 characters

Adds an ldap_full_dn column to store the dn, and only store a sha256
 hash in the ldap_dn which is shorter and can be indexed without
 trouble.
Migration still needs to be implemented.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented on apps/user_ldap/lib/Migration/Version1010Date20200630192842.php:
```diff
@@ -47,7 +47,12 @@ public function changeSchema(IOutput $output, Closure $schemaClosure, array $opt
+				'length' => 64,
+				'default' => '',
+			]);
+			$table->addColumn('ldap_full_dn', Types::STRING, [
```
Can not modify existing migrations. You need a new one

come-nc replied:
Ok, I will move this to a new file

blizzz replied:
`occ migrations:generate` is your friend

come-nc commented:
Note to self: we should look into which kind of hash fits the best what we need here. There are uuid v5, and other hash algorithm, unicity and performance should be compared to select the best fit.

come-nc commented:
Related bug ticket: https://github.com/nextcloud/server/issues/2213

come-nc committed:
Change column names to ldap_dn and ldap_dn_hash and add migration

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen committed:
Fix variable names

Signed-off-by: Joas Schilling <coding@schilljs.com>

come-nc committed:
Fixed migration step for user_ldap

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Put back length check to have a clear error

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fixes in migration step

We cannot set ldap_dn_hash column as notnull because it is empty for
 existing users before postSchemaChange is called

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

blizzz commented:
/rebase

blizzz reviewed (changes requested):
In the first Migration class, the final state of the table should be reflected for new installations. Not necessary to run the while migration history on a new install. So you can add the columns and alter the index in Version1010Date20200630192842.

The new Version1130Date20211102154716 is of course required and looks good as is, with the one remark I left there.

blizzz commented:
@nickvergessen what do you think about its backportability?

blizzz commented:
…while at changing DB it might be worth to add a simple index on directory_uuid, too, since this also may be used as sole WHERE column in queries.

come-nc commented on apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:
```diff
@@ -0,0 +1,138 @@
+					$table->addUniqueIndex(['ldap_dn_hash'], 'ldap_group_dn_hashes');
+					$changeSchema = true;
+				}
+				if (!$table->hasPrimaryKey() || ($table->getPrimaryKeyColumns() !== ['owncloud_name'])) {
```
https://3v4l.org/D9LNm
It seems to work just fine.
And if the function does not return an array it will just evaluate to FALSE.

Not sure where this warning comes from.

nickvergessen commented on apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	protected function handleDNHashes(string $table): void {
+		$q = $this->getSelectQuery($table);
+		$u = $this->getUpdateQuery($table);
```
You don't have to pay per character so for the future value readability over shortening :)

come-nc replied:
Was from a copy/paste of the other migrations, updated both of them in fd43e4e

come-nc committed:
Add the columns and alter the index in Version1010Date20200630192842

This is to ensure new installations do not need to go through migration
 history.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Make sure that hash function returns a string

The documentation says it can return false, and even if that is highly
 unlikely for sha256, better safe than sorry.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add an index for directory_uuid as well

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Use clearer names for variables

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented on apps/user_ldap/lib/Mapping/AbstractMapping.php:
```diff
@@ -178,12 +179,19 @@ public function setUUIDbyDN($uuid, $fdn) {
+	 * Get the hash to store in database column ldap_dn_hash for a given dn
+	 */
+	protected function getDNHash(string $fdn): string {
+		return (string)hash('sha256', $fdn, false);
```
Fixed in f36e11e

come-nc commented:
> In the first Migration class, the final state of the table should be reflected for new installations. Not necessary to run the while migration history on a new install. So you can add the columns and alter the index in Version1010Date20200630192842.

Done in bb6906d

come-nc commented:
> …while at changing DB it might be worth to add a simple index on directory_uuid, too, since this also may be used as sole WHERE column in queries.

Done in c1ff024
Used a unique index as this column should be unique as well, tell me if this needs to be changed to something else.

CarlSchwan commented on apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:
```diff
@@ -0,0 +1,146 @@
+<?php
```
Copyright

skjnldsv commented:
/backport to stable23

skjnldsv commented:
/backport to stable22

skjnldsv commented:
/backport to stable21

skjnldsv commented:
/backport to stable20

CarlSchwan commented:
Sorry for remembering that 2 minutes too late, but I think this is missing a version bump. Otherwise the migration won't run in our development instances.

skjnldsv commented:
@come-nc please do the backports manually :(
